File: 프로그래머스/2/42860. 조이스틱/조이스틱.cpp
#include <string>
#include <vector>
#include <iostream>

using namespace std;

int solution(string name) {
    int ans = 0, n = name.size(), shft = n - 1;
    
    for(int i = 0; i < n; i++) {
        if(name[i] == 'A'){
            int tmp = i;
            while(tmp < n && name[tmp] == 'A') tmp++;
            
            int l = (i == 0) ? 0 : i - 1;
            int r = n - tmp;
            
            shft = min(shft, l + r + min(l,r));
        }
    }
    ans += shft;
    for(auto c: name) ans += min(c - 'A', 'Z' - c + 1);
    return ans;
}

File: 프로그래머스/2/42860. 조이스틱/조이스틱.java
import java.util.*;

class Solution {
    public int solution(String name) {
        int answer = 0, len = name.length(), move = len - 1;
       
        
        for(int i = 0; i < len; i++) {
            answer += Math.min(name.charAt(i) - 'A', Math.abs(name.charAt(i) - 'Z') + 1);
            
            if(i + 1 < len && name.charAt(i + 1) == 'A'){
                int endA = i + 1;
                while(endA < len && name.charAt(endA) == 'A') endA++;
                move = Math.min(move, Math.min(i * 2 + (len - endA), i + (len - endA) * 2));
            }
            
        }
       
        
        return answer + move;
    }
}